### PR TITLE
[feat]: [PL-25416]: Error handling for  bitbucket server in git sync

### PR DESCRIPTION
**Pull request by DeepakPatankar**
Please read [go/trigger-cmd](https://harness.atlassian.net/wiki/spaces/BT/pages/21082441647/PR+Checks) to re-trigger PR Checks using Github comments.

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/33491)

### CODE REVIEW

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerGetDefaultBranchScmApiErrorHandler.java` line 33 at r1](https://reviewable.io/reviews/harness/harness-core/33491#-N2rjnAQ29vhlZnklgmt:-N2rjnAQ29vhlZnklgmu:bw4fqni) ([raw file](https://github.com/harness/harness-core/blob/af115e98422b4a5c4dde96debc21d0c8a28c6dcc/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerGetDefaultBranchScmApiErrorHandler.java#L33)):*
> ```Java
>         throw NestedExceptionUtils.hintWithExplanationException(ScmErrorHints.REPO_NOT_FOUND,
>             GET_DEFAULT_BRANCH_FAILED_MESSAGE + ScmErrorExplanations.REPO_NOT_FOUND,
>             new ScmResourceNotFoundException(errorMessage));
> ```

should be bad request exception

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmFacilitatorServiceImpl.java` line 225 at r1](https://reviewable.io/reviews/harness/harness-core/33491#-N2rk-z15tis-fDEIVok:-N2rk-z15tis-fDEIVol:b1gx56a) ([raw file](https://github.com/harness/harness-core/blob/af115e98422b4a5c4dde96debc21d0c8a28c6dcc/136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmFacilitatorServiceImpl.java#L225)):*
> ```Java
> 
>     if (isFailureResponse(updateFileResponse.getStatus())) {
>       RepoProviders repoProvider =
> ```

@deepakpatankar this is too much redundant work thats being done at every handler. I see that we need url as well to determine the repo provider for bitbucket server. Could you instead take scm connector as argument input in itself in that case and move this repo provider fetching at a central place just like before?

**Review by DeepakPatankar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33491)* status: 35 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmFacilitatorServiceImpl.java` line 225 at r1](https://reviewable.io/reviews/harness/harness-core/33491#-N2rk-z15tis-fDEIVok:-N2tfWRy-mVtE66r1hej:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/af115e98422b4a5c4dde96debc21d0c8a28c6dcc/136-git-sync-manager/src/main/java/io/harness/gitsync/common/impl/ScmFacilitatorServiceImpl.java#L225)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

@deepakpatankar this is too much redundant work thats being done at every handler. I see that we need url as well to determine the repo provider for bitbucket server. Could you instead take scm connector as argument input in itself in that case and move this repo provider fetching at a central place just like before?
</blockquote></details>

Done.
___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerGetDefaultBranchScmApiErrorHandler.java` line 33 at r1](https://reviewable.io/reviews/harness/harness-core/33491#-N2rjnAQ29vhlZnklgmt:-N2tfcrK6GmoOulZNqoF:bhcr3ps) ([raw file](https://github.com/harness/harness-core/blob/af115e98422b4a5c4dde96debc21d0c8a28c6dcc/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerGetDefaultBranchScmApiErrorHandler.java#L33)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

should be bad request exception
</blockquote></details>

For 404 we throw ScmResourceNotFoundException, isn't that correct ?

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerGetDefaultBranchScmApiErrorHandler.java` line 33 at r1](https://reviewable.io/reviews/harness/harness-core/33491#-N2rjnAQ29vhlZnklgmt:-N2ttsp04e6v9gSUKyJo:b-y5sfb) ([raw file](https://github.com/harness/harness-core/blob/af115e98422b4a5c4dde96debc21d0c8a28c6dcc/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerGetDefaultBranchScmApiErrorHandler.java#L33)):*
<details><summary><i>Previously, DeepakPatankar (Deepak Patankar) wrote…</i></summary><blockquote>

For 404 we throw ScmResourceNotFoundException, isn't that correct ?
</blockquote></details>

Idk if you attended the internal discussion where we decided to keep 403, 409, 400 and 500 as valid error codes from SCM service. Issue is 404 is sent out in different cases for different git providers and as an SCM service we can't be consistent in our behaviour in that case. Thus, to maintain consistency, we try to limit our error codes and send out consistent error codes in all different use-cases for different git providers.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreatePullRequestScmApiErrorHandler.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      case 401:
+      case 403:
+        throw NestedExceptionUtils.hintWithExplanationException(ScmErrorHints.INVALID_CREDENTIALS,
+            CREATE_PULL_REQUEST_FAILURE + ScmErrorExplanations.INVALID_CONNECTOR_CREDS,
+            new ScmUnauthorizedException(errorMessage));
+      case 404:
+        throw NestedExceptionUtils.hintWithExplanationException(ScmErrorHints.REPO_NOT_FOUND,
+            CREATE_PULL_REQUEST_FAILURE + ScmErrorExplanations.REPO_NOT_FOUND,
+            new ScmBadRequestException(errorMessage));
+      default:
```
Comment (bhavya181) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreatePullRequestScmApiErrorHandler.java:35]: log the error here

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/exceptions/bitbucketserver/BitbucketServerScmExceptionHints.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+import io.harness.annotations.dev.OwnedBy;
+
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+@OwnedBy(PL)
+public class BitbucketServerScmExceptionHints {
+  public static final String INVALID_CREDENTIALS = "Please check your bitbucket credentials.";
+  public static final String REPO_NOT_FOUND = "Please check your bitbucket repository.";
```
Comment (bhavya181) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/exceptions/bitbucketserver/BitbucketServerScmExceptionHints.java:20]: "Please check if the requested Bitbucket repository exists."

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/exceptions/bitbucketserver/BitbucketServerScmExceptionHints.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import static io.harness.annotations.dev.HarnessTeam.PL;
+
+import io.harness.annotations.dev.OwnedBy;
+
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+@OwnedBy(PL)
+public class BitbucketServerScmExceptionHints {
+  public static final String INVALID_CREDENTIALS = "Please check your bitbucket credentials.";
```
Comment (bhavya181) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/exceptions/bitbucketserver/BitbucketServerScmExceptionHints.java:19]: Please ccheck if your bitbucket creadentials are valid

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerGetFileScmApiErrorHandler.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    switch (statusCode) {
+      case 401:
+      case 403:
+        throw NestedExceptionUtils.hintWithExplanationException(INVALID_CREDENTIALS,
+            GET_FILE_REQUEST_FAILURE + ScmErrorExplanations.INVALID_CONNECTOR_CREDS,
+            new ScmUnauthorizedException(errorMessage));
+      case 404:
+        throw NestedExceptionUtils.hintWithExplanationException(FILE_NOT_FOUND, ScmErrorExplanations.FILE_NOT_FOUND,
+            new ScmBadRequestException(SCMExceptionErrorMessages.FILE_NOT_FOUND_ERROR));
+      default:
```
Comment (bhavya181) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerGetFileScmApiErrorHandler.java:33]: log the error here

**Review by bhavya181 — CHANGES_REQUESTED** [review]
> Reviewed 12 of 35 files at r1, 11 of 23 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33491)* status: 44 file(s) to review. 1 change requests. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreateFileScmApiErrorHandler.java` line 37 at r3](https://reviewable.io/reviews/harness/harness-core/33491#-N2vDEoeEdKP0Klv8oAZ:-N2vDEoeEdKP0Klv8oA_:b-pl1he4) ([raw file](https://github.com/harness/harness-core/blob/13951a9b3ce46d01f01ada19d13a27086e544002/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreateFileScmApiErrorHandler.java#L37)):*
> ```Java
>         throw NestedExceptionUtils.hintWithExplanationException(ScmErrorHints.FILE_ALREADY_EXISTS,
>             CREATE_FILE_REQUEST_FAILURE + ScmErrorExplanations.FILE_ALREADY_EXISTS,
>             new ScmBadRequestException(errorMessage));
> ```

Here ScmConflictError should be thrown

**Review by mohitg0795 — DISMISSED** [review]
:lgtm:

> Reviewed 21 of 35 files at r1, 13 of 23 files at r3, 10 of 10 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33491#-:-N2zBL_QDaFPvMoAX2jw:bnfp4nl)* status: 29 file(s) to review. 1 change requests. 3. unresolved discussions. 1 of 2 approvals obtained.

**Comment by DeepakPatankar** [comment]
trigger messagecheck

**Review by ujjawal-harness — DISMISSED** [review]
:lgtm:

> Reviewed 21 of 35 files at r1, 13 of 23 files at r3, 10 of 10 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33491#-:-N2zTd4pBifaXwGMxdlD:bnfp4nl)* status: 1 change requests. 3. unresolved discussions.

**Review by meenakshi-harness — APPROVED** [review]
> Reviewed 21 of 35 files at r1, 13 of 23 files at r3, 10 of 10 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33491)* status: 1 change requests. 3. unresolved discussions.

**Review by bhavya181 — COMMENTED** [review]
> Reviewed 9 of 35 files at r1, 10 of 23 files at r3, 10 of 10 files at r4, 3 of 5 files at r5.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33491)* status: 5 file(s) to review. 1 change requests. 

___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreateFileScmApiErrorHandler.java` line 37 at r3](https://reviewable.io/reviews/harness/harness-core/33491#-N2vDEoeEdKP0Klv8oAZ:-N3-EnhZ-5RLjrQpdf9q:bp39sau) ([raw file](https://github.com/harness/harness-core/blob/13951a9b3ce46d01f01ada19d13a27086e544002/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreateFileScmApiErrorHandler.java#L37)):*
<details><summary><i>Previously, DeepakPatankar (Deepak Patankar) wrote…</i></summary><blockquote>

V2 one ?
</blockquote></details>

Sorry, we have decided to keep scmbadRequest here and only in case of update where conflicts occur we will send scmConflict exception. Please change it.
___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreatePullRequestScmApiErrorHandler.java` line 35 at r2](https://reviewable.io/reviews/harness/harness-core/33491#-N2pd3gP2_5RsYK1OtWP-r2-35:-N3-F2C29MXxSJS8RsTM:bvrg8uk) ([raw file](https://github.com/harness/harness-core/blob/fae912bc893ae360ebe6fff13e86acd13811ecd0/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreatePullRequestScmApiErrorHandler.java#L35)):*
<details><summary><i>Previously, DeepakPatankar (Deepak Patankar) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Please ceck there are many places where logging is not done in case of default.
___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerGetFileScmApiErrorHandler.java` line 33 at r2](https://reviewable.io/reviews/harness/harness-core/33491#-N2pd3gP2_5RsYK1OtWR-r2-33:-N3-FBEIBWeH0EUi_0l6:bb7od0r) ([raw file](https://github.com/harness/harness-core/blob/fae912bc893ae360ebe6fff13e86acd13811ecd0/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerGetFileScmApiErrorHandler.java#L33)):*
<details><summary><i>Previously, DeepakPatankar (Deepak Patankar) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

same

**Review by DeepakPatankar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33491)* status: 7 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreateFileScmApiErrorHandler.java` line 37 at r3](https://reviewable.io/reviews/harness/harness-core/33491#-N2vDEoeEdKP0Klv8oAZ:-N3-GWQN6-a9CqKVAnLd:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/13951a9b3ce46d01f01ada19d13a27086e544002/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreateFileScmApiErrorHandler.java#L37)):*
<details><summary><i>Previously, bhavya181 (Bhavya Agrawal) wrote…</i></summary><blockquote>

Sorry, we have decided to keep scmbadRequest here and only in case of update where conflicts occur we will send scmConflict exception. Please change it.
</blockquote></details>

Done.
___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreatePullRequestScmApiErrorHandler.java` line 35 at r2](https://reviewable.io/reviews/harness/harness-core/33491#-N2pd3gP2_5RsYK1OtWP-r2-35:-N3-GYHJ1yVFxQZl1vRy:b-5cxb7z) ([raw file](https://github.com/harness/harness-core/blob/fae912bc893ae360ebe6fff13e86acd13811ecd0/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreatePullRequestScmApiErrorHandler.java#L35)):*
<details><summary><i>Previously, bhavya181 (Bhavya Agrawal) wrote…</i></summary><blockquote>

Please ceck there are many places where logging is not done in case of default.
</blockquote></details>

Added everywhere
___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerGetFileScmApiErrorHandler.java` line 33 at r2](https://reviewable.io/reviews/harness/harness-core/33491#-N2pd3gP2_5RsYK1OtWR-r2-33:-N3-GZtLBB-fSEfIvPBq:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/fae912bc893ae360ebe6fff13e86acd13811ecd0/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerGetFileScmApiErrorHandler.java#L33)):*
<details><summary><i>Previously, bhavya181 (Bhavya Agrawal) wrote…</i></summary><blockquote>

same
</blockquote></details>

Done.

**Comment by DeepakPatankar** [comment]
trigger ti4

**Comment by DeepakPatankar** [comment]
trigger ti1

**Review by bhavya181 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33491#-:-N3-gAt2-OE_wSJ1c5BP:bnfp4nl)* status: 7 file(s) to review. 1 of 2 approvals obtained.

**Review by mohitg0795 — APPROVED** [review]
:lgtm:

> Reviewed 5 of 5 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33491#-:-N3-l-Kd3CEb5iTGzSAg:bnfp4nl)* status: :shipit: complete! ✓

**Pull request merged**